File: app/src/main/java/com/sysu/edu/browser/JSEditorActivity.java
package com.sysu.edu.browser;

import android.content.ContentValues;
import android.os.Bundle;

import com.alibaba.fastjson2.JSONObject;
import com.sysu.edu.BaseActivity;
import com.sysu.edu.R;
import com.sysu.edu.databinding.ActivityJsEdiitorBinding;

import org.eclipse.tm4e.core.registry.IThemeSource;

import java.util.Objects;

import io.github.rosemoe.sora.langs.textmate.TextMateColorScheme;
import io.github.rosemoe.sora.langs.textmate.TextMateLanguage;
import io.github.rosemoe.sora.langs.textmate.registry.FileProviderRegistry;
import io.github.rosemoe.sora.langs.textmate.registry.GrammarRegistry;
import io.github.rosemoe.sora.langs.textmate.registry.ThemeRegistry;
import io.github.rosemoe.sora.langs.textmate.registry.model.ThemeModel;
import io.github.rosemoe.sora.langs.textmate.registry.provider.AssetsFileResolver;

public class JSEditorActivity extends BaseActivity {
    
    BrowserHelper db;
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityJsEdiitorBinding binding = ActivityJsEdiitorBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        JSONObject data = JSONObject.parseObject(getIntent().getStringExtra("item"));
        if (data != null) {
            db = new BrowserHelper(this);
            FileProviderRegistry.getInstance().addFileProvider(new AssetsFileResolver(getAssets()));
            GrammarRegistry.getInstance().loadGrammars("textmate/languages.json");
            binding.editor.setEditorLanguage(TextMateLanguage.create("source.js", true));
            var themeRegistry = ThemeRegistry.getInstance();
            var name = "light"; // 主题名称
            var themeAssetsPath = "textmate/" + name + ".json";
            var model = new ThemeModel(IThemeSource.fromInputStream(Objects.requireNonNull(FileProviderRegistry.getInstance().tryGetInputStream(themeAssetsPath)), themeAssetsPath, null), name);
            try {
                themeRegistry.loadTheme(model);
                themeRegistry.setTheme(name);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
            binding.editor.setColorScheme(TextMateColorScheme.create(themeRegistry));
            getMenuInflater().inflate(R.menu.editor, binding.toolbar.getMenu());
            binding.toolbar.setNavigationOnClickListener(_ -> supportFinishAfterTransition());
            var value = new ContentValues();
            binding.toolbar.setOnMenuItemClickListener(item -> {
                if (item.getItemId() == R.id.save) {
                    value.clear();
                    value.put("script", binding.editor.getText().toString());
                    db.getWritableDatabase().update("js", value, "id = ?", new String[]{data.getString("id")});
                } else if (item.getItemId() == R.id.redo) binding.editor.redo();
                else if (item.getItemId() == R.id.undo) binding.editor.undo();
                return false;
            });
            binding.editor.setText(data.getString("script"));
        }
    }
}

File: app/src/main/java/com/sysu/edu/browser/JSEditorActivity.kt
package com.sysu.edu.browser

import android.os.Bundle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import com.sysu.edu.BaseActivity
import com.sysu.edu.R
import com.sysu.edu.browser.data.BrowserRepository
import com.sysu.edu.browser.data.JavaScriptEntity
import com.sysu.edu.browser.data.JsModel
import com.sysu.edu.browser.data.JsModelFactory
import com.sysu.edu.browser.data.ScriptParser.updateEntityByScript
import com.sysu.edu.databinding.ActivityJsEdiitorBinding
import io.github.rosemoe.sora.langs.textmate.TextMateColorScheme
import io.github.rosemoe.sora.langs.textmate.TextMateLanguage
import io.github.rosemoe.sora.langs.textmate.registry.FileProviderRegistry
import io.github.rosemoe.sora.langs.textmate.registry.GrammarRegistry
import io.github.rosemoe.sora.langs.textmate.registry.ThemeRegistry
import io.github.rosemoe.sora.langs.textmate.registry.model.ThemeModel
import io.github.rosemoe.sora.langs.textmate.registry.provider.AssetsFileResolver
import org.eclipse.tm4e.core.registry.IThemeSource

class JSEditorActivity : BaseActivity() {
	val model: JsModel by lazy {
		ViewModelProvider(this,
		                  JsModelFactory(BrowserRepository(this,
		                                                   lifecycleScope)))[JsModel::class.java]
	}
	
	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		val binding = ActivityJsEdiitorBinding.inflate(layoutInflater)
		setContentView(binding.root)
		model.getJs(intent.getLongExtra("id", 0)) { data ->
			if (data != null) {
				var data: JavaScriptEntity = data
				FileProviderRegistry.getInstance().addFileProvider(AssetsFileResolver(assets))
				GrammarRegistry.getInstance().loadGrammars("textmate/languages.json")
				binding.editor.setEditorLanguage(TextMateLanguage.create("source.js", true))
				val themeRegistry = ThemeRegistry.getInstance()
				val name = "light" // 主题名称
				val themeAssetsPath = "textmate/$name.json"
				val themeModel = ThemeModel(IThemeSource.fromInputStream(FileProviderRegistry.getInstance()
					                                                         .tryGetInputStream(
						                                                         themeAssetsPath),
				                                                         themeAssetsPath,
				                                                         null), name)
				try {
					themeRegistry.loadTheme(themeModel)
					themeRegistry.setTheme(name)
				} catch (e: Exception) {
					throw RuntimeException(e)
				}
				binding.editor.setColorScheme(TextMateColorScheme.create(themeRegistry))
				menuInflater.inflate(R.menu.editor, binding.toolbar.menu)
				binding.toolbar.setNavigationOnClickListener { supportFinishAfterTransition() }
				binding.toolbar.setOnMenuItemClickListener {
					when (it.itemId) {
						R.id.save -> {
							data = updateEntityByScript(data, binding.editor.text.toString())
							model.updateJs(data)
						}
						R.id.redo -> binding.editor.redo()
						R.id.undo -> binding.editor.undo()
					}
					false
				}
				binding.editor.setText(data.script)
			}
		}
	}
}